src/details/ArborX_MinimumSpanningTree.hpp

```diff
@@ -504,6 +503,142 @@ void finalizeEdges(ExecutionSpace const &space, BVH const &bvh,
+          sided_parents(e) = FOLLOW_CHAIN_VALUE - alpha_edge_index;
+        }
+      });
+}
+
+template <class ExecutionSpace, class Labels, class OutEdges,
+          class EdgesMapping, class BVH, class Parents>
+void assignVertexParents(ExecutionSpace const &space, Labels const &labels,
+                         OutEdges const &out_edges,
+                         EdgesMapping const &edges_mapping, BVH const &bvh,
+                         Parents parents)
+{
+  auto const n = edges_mapping.extent_int(0) + 1;
+  int const vertices_offset = n - 1;
+
+  Kokkos::parallel_for(
+      "ArborX::MST::compute_vertex_parents",
+      Kokkos::RangePolicy<ExecutionSpace>(space, 0, n), KOKKOS_LAMBDA(int e) {
+        auto const &edge = out_edges(e);
+
+        int i = labels(edge.source());
+        parents(HappyTreeFriends::getLeafPermutationIndex(bvh, i) +
+                vertices_offset) = edges_mapping(i - vertices_offset);
+      });
+}
+
+template <typename ExecutionSpace, typename Edges, typename SidedParents,
+          typename Parents>
+void computeParents(ExecutionSpace const &space, Edges const &edges,
+                    SidedParents const &sided_parents, Parents &parents)
+{
+  KokkosExt::ScopedProfileRegion guard("ArborX::MST::compute_edge_parents");
+
+  using MemorySpace = typename SidedParents::memory_space;
+
+  auto num_edges = edges.size();
+
+  // Encode both sided parent and edge into long long
+  // This way, once we sort based on this value, edges with the same sided
+  // parent will already be sorted in increasing order.
+  Kokkos::View<long long *, MemorySpace> keys(
+      Kokkos::view_alloc(space, Kokkos::WithoutInitializing,
+                         "ArborX::MST::keys"),
+      num_edges);
+
+  constexpr int shift = 32;
+
+  union Pun
+  {
+    float f;
+    int i;
+  };
+
+  Kokkos::parallel_for(
+      "ArborX::MST::compute_sided_alpha_parents",
+      Kokkos::RangePolicy<ExecutionSpace>(space, 0, num_edges),
+      KOKKOS_LAMBDA(int const e) {
+        long long key = sided_parents(e);
+        auto const &edge = edges(e);
+        if (key <= FOLLOW_CHAIN_VALUE)
+        {
+          int next = FOLLOW_CHAIN_VALUE - key;
+          do
+          {
+            key = sided_parents(next);
+            if (key <= FOLLOW_CHAIN_VALUE)
+              next = FOLLOW_CHAIN_VALUE - key;
+            else if (key >= 0)
+            {
+              next = key / 2;
+              auto const &next_edge = edges(next);
+              if (edge < next_edge)
+                break;
+            }
+            else if (key == ROOT_CHAIN_VALUE)
+              break;
+          } while (true);
+        }
+        if (key == ROOT_CHAIN_VALUE)
+          key = INT_MAX;
+
+        // Comparison of weights as ints is the same as their comparison as
+        // floats as long as they are positive and are not NaNs or inf
+        assert(sizeof(int) == sizeof(float));
+        Pun p = {edge.weight};
+        keys(e) = (key << shift) + p.i;
```

Comment:
Any reason not to just cast `edge.weight` to `int`? This looks pretty sketchy.

Reply:
Since you are not just manipulating bits here, the use of a union in such a way is invoked behavior (but supported by a lot of compilers). Note that you could also use 
```
reinterpret_cast<int const&>(edge.weight);
```
but really should use
```
std::bit_cast<int>(edge.weight);
```
(if only that was portable and available before C++20...). I'm not that comfortable with `float1< float2` implying `std::bit_cast<int>(float1) < std::bit_cast<int>(float2);` but I guess that's reasonably safe.

Reply:
> `reinterpret_cast<int const&>(edge.weight);`

This won't compile, as it need pointers. And using reinterpret_cast on pointers won't compiler either resulting in illegal type punning errror.

> `std::bit_cast<int>(edge.weight);`

bit_cast is C++20, so unless I use a Kokkos routine to do that, it's unavailable.

> I'm not that comfortable with float1< float2 implying std::bit_cast<int>(float1) < std::bit_cast<int>(float2); but I guess that's reasonably safe.

Neither am I, but it seems correct and the only way I see right now to avoid a more expensive sort.

Reply:
For completeness, `reinterpret_cast` should really work here, see https://godbolt.org/z/heq6fq63K (both type punning approaches do the exact same thing for -O1 and higher).